Buffered forwarder for #15

laher opened the pull request:
@peterbourgon please see this experimental implementation for #15 , although it's _a bit racey_ for the time being. Please don't consider this complete yet.

 * It basically works - when I kill my `ingeststore` the forwarder buffers some messages, and then after restarting the ingeststore it reconnects & forwards the buffer :+1:  ... *but* sometimes a couple of messages get sent twice - it needs work still.
 * I've been testing it by piping `date` once per second onto `oklog forward -buf -bufsize=5 localhost`, then repeatedly querying `oklog query -from 30s` in another window, and then killing/restarting my ingeststore for 5-15 seconds. It's pretty easy to see what's going on.
 * For the buffer I tried using `container/ring` from the standard library. I added a mutex and a couple of other fields to maintain state. Seems OKish, but not very simple really. Any advice on how you envisioned this?
 * I chose messages as the unit for buffering, rather than bytes. I figured it fits the forwarding code reasonably well.

It's late now so I just thought I'd elicit some feedback for now, as you may have had something very different in mind.
Cheers

laher commented:
OK did a bit more this morning, and I'm happier with the approach now - some work still needed, but it's working nicely and the approach feels more sane. The ring buffer now implements a `textScanner` interface, to act as a drop-in replacement for the `bufio.Scanner`.
 
I'll report back once I've got it 100%.

peterbourgon commented on cmd/oklog/forward_buffer.go:
```diff
@@ -0,0 +1,83 @@
+
+func (bf *ringBuffer) Scan() bool {
+	return true
+}
```
As noted above, we're gonna want to propagate a shutdown signal thru somehow to terminate this.

peterbourgon commented on cmd/oklog/forward_buffer.go:
```diff
@@ -0,0 +1,83 @@
+	bf.r = bf.r.Next()
+	bf.inc(-1)
+	return v
+}
```
Unless I'm really misunderstanding something, this method has a number of problems…

- It's burning the CPU calling bf.Remaining over and over
- Once that loop exits, the bf.r.Value is not guaranteed to be there, another goroutine calling Get could have stolen the work item — a data race
- bf.r is written without synchronization — a data race

I think using a `remaining` counter as the synchronization point is not really viable, at least not as implemented here...

laher replied:
Agreed. It's not great. I'm not loving `container/ring` TBH. I'd like to grok it better with some more investigation into the Link/Unlink methods, but I'm also thinking about alternatives.

I'd like to try a buffered channel implementation (2 channels, similar to this https://content.pivotal.io/blog/a-channel-based-ring-buffer-in-go ). It would be safe from races but I'm not sure how performance would compare to locking approaches. I guess a benchmark would help confirm or otherwise.

peterbourgon replied:
Check the comments on that Pivotal buffer, it's racy as printed.

laher replied:
OK, I'll see if I can make container/ring work without the counter and then look at other options from there. Cheers

laher commented on cmd/oklog/forward.go:
```diff
@@ -21,8 +21,10 @@ import (
+		apiAddr     = flagset.String("api", "", "listen address for forward API (and metrics)")
+		prefixes    = stringslice{}
+		ringBuf     = flagset.Bool("buf", false, "use a ring buffer to drop old messages when e.g. the ingestion engine fails")
+		ringBufSize = flagset.Int("bufsize", 1024, "ring buffer size (messages)")
```
Yup, sure. Sounds like you'd like to keep options open for other alternatives in future.

laher commented:
I've applied the straightforward changes, thanks for your guidance. I'll work on the ring buffer itself in due course and add appropriate tests/benchmarks. Cheers

laher commented:
@peterbourgon I've rewritten the buffer now, so that `Get()` blocks on a channel whilst data is unavailable.

So, I tried starting from scratch: 2 new struct types, composed together with an interface, and backed by a slice and a channel.
 
 * `bufferedScanner` is forwarding messages to a 'buffer' type, implementing some methods which are shared with `io.Scanner`.
 * `sliceBuffer` is the only implementation that I've added. 
 * `sliceBuffer.Get()` blocks on a channel whenever the buffer is empty, and no data is available.
 * `sliceBuffer.Put()` attempts to deliver messages on the channel. If the channel is not being listened to, Put adds the messages onto a bounded buffer (a slice). Once the maximum size is reached, further messages are discarded.

Hoping this change is closer to the mark. 
Thanks

laher commented on cmd/oklog/forward_slicebuffer.go:
```diff
@@ -0,0 +1,67 @@
+func (b *sliceBuffer) Put(record string) {
+	select {
+	case b.ch <- record:
+	default:
```
I've used the select/default on the send side only, for one-way blocking. We always want Get to wait for data, but we want Put to buffer instead.

laher commented on cmd/oklog/forward_slicebuffer.go:
```diff
@@ -0,0 +1,67 @@
+	}
+	b.mutex.RUnlock()
+	b.mutex.Lock()
+	record, b.buf = b.buf[0], b.buf[1:]
```
Hmm, I think maybe the best bet is to keep it as a slice but to use it like a ring-buffer to reduce the number of allocations. I'll have a play with it and try some `-memprofile` analysis

laher commented:
Please see #62 for a proposed direction I'd like to take this feature.

(If 62 gets rejected, perhaps we can still come back to this)

laher commented:
OK I think I'm done with this buffered-channel implementation of the ring buffer. I think it's nice and simple.

I fancy you'll have some more thoughts on the `bufferedScanner`. I have amended the API a little to make it a little simpler IMO. Just one call to Next() instead of Scan()/Text()/Error(). It makes more sense to me.

Still, I'm good for now. I'll go over and close the other PR. Cheers.

laher commented:
@peterbourgon have you had a chance to look at the most recent implementation with the buffered channel?

Ta

peterbourgon commented:
Not yet, sorry! I will find some time on the weekend 🤞
